### DISCUSSION TITLE
How to understand points in raster cell center

### DISCUSSION THREAD

**Discussion by zhengliuX** [pulseq/pulseq#42] [Q&A] [answered]
Hi, Dear @pulseq community, I've got some question about raster time. When reading specification section 2.6, it says 
```
In contrast to that, any default sampling points (e.g. sampling points of arbitrary-shape gradients) are aligned to the centers of the corresponding raster steps.
```
Which means for **RF**, **Arbitrary Gradient** and **ADC**, when you are designing the these events, I should know that the points I designed will be in the center of a raster time cell.  

I've read https://github.com/pulseq/pulseq/issues/32#issuecomment-1707839354, if I understand that discussion right, when designing an ADC event, the real points in the center of raster time cell are the accurate I want the points to be, no matter how the real hardware designs, as for hardware may need some warm-up time, causing a little delay. Is this the same way to understand my RF and arbitrary gradients points? The accurate time I want the points generated by hardware. 

So if I set RF delay as rf_delay, **the first point** should be 
```
rf_delay + 0.5 * rf_dwelltime
```
same for arbitrary gradients 
```
grad_delay + 0.5 * grad_dwelltime
```
and that is the accurate time the point will be. To do a tricky thing, if I want events to start at rf_delay(for RF), in my code, I should set delay as 
```
rf_delay - 0.5 * rf_dwelltime
```
Then Pulseq will set it to 
```
rf_delay - 0.5 * rf_dwelltime + 0.5 * rf_dwelltime = rf_delay
```

To a Pulseq interpreter, when parsing a .seq file, it gets the accurate time of the points. Let's say it gets the first point start time is `rf_delay + 0.5 * rf_dwelltime`, which is calculated using `rf.delay + rf.t[0]`, then the interpreter just makes sure the first point occurring at `rf_delay + 0.5 * rf_dwelltime` second relative to the start time of the block.

--- 

Also, for hard RF pulseq(or block pulseq based on https://github.com/pulseq/pulseq/blob/master/matlab/%2Bmr/makeBlockPulse.m#L57), 
it's defined on the edge, different from other RF pulses, is this excepted?

Thank a lot!
Zheng

**Comment by pulseq-admin** [comment]
Dear Zheng, 

sorry, we've missed your question, so we reopen it now, because this is an important point. 

There are two types of shapes in Pulseq: 

1: idealized polygon shapes: these consist of explicit time points and values at these time points. The values in between are restored by linear interpolation; 
2: explicitly sampled shapes. 

Example of the first type of the shape include normal trapezoid gradients, extended trapezoids or block RF pulses. Examples of the second type include arbitrary gradients and arbitrary RF pulses. Also the sampled MRI data (as recorded by the ADC directives) belong to the shape type 2. 

The shape type 1 defines the time points on the raster edges. The shape type 2 defines the raster points on the raster centers. 

Below are some slides, which may make the related concepts more clear: 

![image](https://github.com/pulseq/pulseq/assets/12132792/4c1d7c5c-73d2-40d0-8ee1-2f7b1788dc9b)
![image](https://github.com/pulseq/pulseq/assets/12132792/53bdd019-9221-48b8-9795-ed12522927ab)
![image](https://github.com/pulseq/pulseq/assets/12132792/4a750570-75f9-41c1-be23-f2c744fbacd3)
![image](https://github.com/pulseq/pulseq/assets/12132792/9222d3fb-355a-4bb8-b74c-314c5df92bd9)

Hope this helps!
--Maxim

**Comment by pulseq-admin** [comment, marked as answer]
If you implement your own interpreter for a particular hardware, in case your hardware uses a different conventions, you have to add delays to different types of objects in a way that is similar to what you describe in your question. These delays should be hissed from the end users as along as you can guarantee, that your interpreter is able of following the Pulseq raster/sampling convention.
